File: TheaterSuggestions/CSharp/SeatsSuggestions.Domain/SeatAllocator.cs
using System.Collections.Generic;
using System.Threading.Tasks;
using SeatsSuggestions.Domain.Port;

namespace SeatsSuggestions.Domain
{
    public class SeatAllocator : IProvideAuditoriumSeating
    {
        private const int NumberOfSuggestionsPerPricingCategory = 3;
        private readonly IRetrieveAuditoriumSeating _retrieveAuditoriumSeating;

        public async Task<SuggestionsMade> MakeSuggestions(ShowId showId, PartyRequested partyRequested)
        {
            var suggestionsMade = new SuggestionsMade(showId, partyRequested);

            var auditoriumSeating = await _retrieveAuditoriumSeating.GetById(showId);

            foreach (var pricingCategory in new List<PricingCategory> { PricingCategory.First, PricingCategory.Second, 
                                                                        PricingCategory.Third, PricingCategory.Mixed })
            {
                suggestionsMade.Add(GiveMeSuggestionsFor(auditoriumSeating, partyRequested, pricingCategory));
            }

            _retrieveAuditoriumSeating.Save(auditoriumSeating);

            return suggestionsMade.MatchExpectations() ? suggestionsMade : new SuggestionNotAvailable(showId, partyRequested);
        }

        public SeatAllocator(IRetrieveAuditoriumSeating retrieveAuditoriumSeating)
        {
            _retrieveAuditoriumSeating = retrieveAuditoriumSeating;
        }

        private static IEnumerable<SuggestionMade> GiveMeSuggestionsFor(
            AuditoriumSeating auditoriumSeating,
            PartyRequested partyRequested,
            PricingCategory pricingCategory)
        {
            var foundedSuggestions = new List<SuggestionMade>();

            for (var i = 0; i < NumberOfSuggestionsPerPricingCategory; i++)
            {
                var seatOptionsSuggested = auditoriumSeating
                    .SuggestSeatingOptionFor(new SuggestionRequest(partyRequested, pricingCategory));

                if (seatOptionsSuggested.MatchExpectation())
                {
                    // We get the new version of the Auditorium after the allocation
                    auditoriumSeating = auditoriumSeating.Allocate(seatOptionsSuggested);

                    foundedSuggestions.Add(new SuggestionMade(partyRequested, pricingCategory,
                        seatOptionsSuggested.Seats));
                }
            }

            return foundedSuggestions;
        }
    }
}

File: TheaterSuggestions/Java/SeatsSuggestionsDomain/src/main/java/com/baasie/SeatsSuggestionsDomain/SeatAllocator.java
package com.baasie.SeatsSuggestionsDomain;

import com.baasie.SeatsSuggestionsDomain.port.IAdaptAuditoriumSeating;
import com.baasie.SeatsSuggestionsDomain.port.IProvideAuditoriumSeating;
import com.google.common.collect.ImmutableList;

import java.util.ArrayList;
import java.util.List;

public class SeatAllocator implements IProvideAuditoriumSeating {
    private static final int NUMBER_OF_SUGGESTIONS = 3;
    private final IAdaptAuditoriumSeating iAdaptAuditoriumSeating;

    public SeatAllocator(IAdaptAuditoriumSeating iAdaptAuditoriumSeating) {
        this.iAdaptAuditoriumSeating = iAdaptAuditoriumSeating;
    }

    private static List<SuggestionMade> giveMeSuggestionsFor(
            AuditoriumSeating auditoriumSeating, PartyRequested partyRequested, PricingCategory pricingCategory) {

        var suggestionRequest = new SuggestionRequest(partyRequested, pricingCategory);
        List<SuggestionMade> foundedSuggestions = new ArrayList<>();
        for (var i = 0; i < NUMBER_OF_SUGGESTIONS; i++) {
            SeatingOptionSuggested seatingOptionSuggested = auditoriumSeating.suggestSeatingOptionFor(suggestionRequest);

            if (seatingOptionSuggested.matchExpectation()) {
                // We get the new version of the Auditorium after the allocation
                auditoriumSeating = auditoriumSeating.allocate(seatingOptionSuggested);
                foundedSuggestions.add(new SuggestionMade(seatingOptionSuggested.seats(), partyRequested, pricingCategory));
            }
        }

        return ImmutableList.copyOf(foundedSuggestions);
    }

    public SuggestionsMade makeSuggestions(ShowId showId, PartyRequested partyRequested) {

        var auditoriumSeating = iAdaptAuditoriumSeating.getAuditoriumSeating(showId);

        var suggestionsMade = new SuggestionsMade(showId, partyRequested);

        suggestionsMade.add(giveMeSuggestionsFor(auditoriumSeating, partyRequested,
                PricingCategory.First));
        suggestionsMade.add(giveMeSuggestionsFor(auditoriumSeating, partyRequested,
                PricingCategory.Second));

        suggestionsMade.add(giveMeSuggestionsFor(auditoriumSeating, partyRequested,
                PricingCategory.Third));
        suggestionsMade.add(giveMeSuggestionsFor(auditoriumSeating, partyRequested,
                PricingCategory.Mixed));

        if (suggestionsMade.matchExpectations()) {
            return suggestionsMade;
        }

        return new SuggestionNotAvailable(showId, partyRequested);
    }
}